expression/builtin_regexp.go

```diff
@@ -0,0 +1,1400 @@
+// within match_type, the rightmost one takes precedence.
+func (re *regexpBaseFuncSig) getMatchType(userInputMatchType string) (string, error) {
+	flag := ""
+	matchTypeSet := set.NewStringSet()
+
+	if collate.IsCICollation(re.baseBuiltinFunc.collation) {
+		matchTypeSet.Insert(flagI)
+	}
+
+	for _, val := range userInputMatchType {
+		c := string(val)
+
+		// Check validation of the flag
+		_, err := validMatchType[c]
+		if !err {
+			return "", ErrRegexp.GenWithStackByArgs(invalidMatchType)
+		}
+
+		if c == flagC {
+			// re2 is case-sensitive by default, so we only need to delete 'i' flag
+			// to enable the case-sensitive for the regexp
+			delete(matchTypeSet, flagI)
+			continue
+		}
+
+		matchTypeSet[c] = empty{} // add this flag
+	}
+
+	// generate flag
+	for key := range matchTypeSet {
+		flag += key
+	}
+
+	return flag, nil
+}
+
+// To get a unified compile interface in initMemoizedRegexp, we need to process many things in genCompile
+func (re *regexpBaseFuncSig) genCompile(matchType string) (func(string) (*regexp.Regexp, error), error) {
+	matchType, err := re.getMatchType(matchType)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(pat string) (*regexp.Regexp, error) {
+		if len(matchType) == 0 {
+			return regexp.Compile(pat)
+		}
+		return regexp.Compile(fmt.Sprintf("(?%s)%s", matchType, pat))
+	}, nil
+}
+
+func (re *regexpBaseFuncSig) genRegexp(pat string, matchType string) (*regexp.Regexp, error) {
+	if re.isMemorizedRegexpInitialized() {
+		return re.memorizedRegexp, re.memorizedErr
+	}
+
+	var err error
+
+	// Generate compiler first
+	compile, err := re.genCompile(matchType)
+	if err != nil {
+		return nil, err
+	}
+
+	return compile(pat)
+}
+
+// we can memorize the regexp when:
+//  1. pattern and match type are constant
+//  2. pattern is const and there is no match type argument
+//
+// return true: need, false: needless
+func (re *regexpBaseFuncSig) canMemorize(matchTypeIdx int) bool {
+	return re.args[patternIdx].ConstItem(re.ctx.GetSessionVars().StmtCtx) && (len(re.args) <= matchTypeIdx || re.args[matchTypeIdx].ConstItem(re.ctx.GetSessionVars().StmtCtx))
+}
+
+func (re *regexpBaseFuncSig) initMemoizedRegexp(params []*regexpParam, matchTypeIdx int) error {
+	// Generate compile
+	compile, err := re.genCompile(params[matchTypeIdx].getStringVal(0))
+	if err != nil {
+		return ErrRegexp.GenWithStackByArgs(err)
+	}
+
+	// Compile this constant pattern, so that we can avoid this repeatable work
+	re.memorize(compile, params[patternIdx].getStringVal(0))
+
+	return re.memorizedErr
+}
+
+// As multiple threads may memorize regexp and cause data race, only the first thread
+// who gets the lock is permitted to do the memorization and others should wait for him
+// until the memorization has been finished.
+func (re *regexpBaseFuncSig) tryToMemorize(params []*regexpParam, matchTypeIdx int, n int) error {
+	// Check memorization
+	if n == 0 || !re.canMemorize(matchTypeIdx) {
```

Comment:
The `n` here is just to make sure if the memorization is necessary.
How about moving this judgement out of this function?

Reply:
> 

`try` in the function name means this function will check all the condition and user is needless to care about these judgements.